### PR TITLE
Draft - Android - allow supply a different request factory to HttpClient

### PR DESCRIPTION
**Pull request by anod**
Task #1049 Enable HTTP/2.0 support for the SDK.

- Add option to provide a different request factory to HttpClient
   - Add `HttpClientRequest` and `HttpClientRequest.Factory` interfaces
   - Supply a default `AndroidUrlConnection` factory
- Add a new `maesdk-okhttp` module with okhttp implementation to avoid pulling/managing external dependencies

image

TODO:
[  ] Test okhttp: parsing request, cancellation, http/2 connection are being reuse
[  ] Add unit test for okhttp
[  ] Publish okhttp module

### CODE REVIEW

File: lib/android_build/app/build.gradle [hunk 1]
```diff
@@ -30,14 +30,19 @@ android {
...
+    compileOptions {
+        sourceCompatibility JavaVersion.VERSION_1_8
+        targetCompatibility JavaVersion.VERSION_1_8
+    }
 }
 
 dependencies {
     implementation fileTree(dir: 'libs', include: ['*.jar'])
     implementation project(':maesdk')
+    implementation project(':maesdk-okhttp')
```
Comment (sid-dahiya) [inline lib/android_build/app/build.gradle:42]: What is `okhttp`?

**Comment by sid-dahiya** [comment]
Why don't we have okhttp implementation be the default implementation? @lalitb ?

**Comment by anod** [comment]
> Why don't we have okhttp implementation be the default implementation? @lalitb ?

need to cosider version management: default will require all clients to update to 4.10, and vise versa in case of new version of okhttp, the library needs to be updated...

**Comment by sid-dahiya** [comment]
> > Why don't we have okhttp implementation be the default implementation? @lalitb ?
> 
> need to cosider version management: default will require all clients to update to 4.10, and vise versa in case of new version of okhttp, the library needs to be updated...

Got it, in that case, maybe this can be the default in the next major SDK version.

File: lib/android_build/app/build.gradle [hunk 2]
```diff
@@ -30,14 +30,19 @@ android {
             version "3.10.2"
         }
     }
+    compileOptions {
+        sourceCompatibility JavaVersion.VERSION_1_8
+        targetCompatibility JavaVersion.VERSION_1_8
+    }
```
Comment (lalitb) [inline lib/android_build/app/build.gradle:36]: Is this dependency coming for okhttp ? As I understand there are small number of android devices on java 6/7, the build will fail for those?

Comment (anod) [reply lib/android_build/app/build.gradle:36]: latest android stuido uses java 11 by default, so it failed to compile, this tells it use 1.8 like in maesdk library

**Comment by lalitb** [comment]
@sid-dahiya  Do you know if someone from office telemetry team with Android expertise (Saurabh?) can help review this code?

**Comment by anod** [comment]
> > > Why don't we have okhttp implementation be the default implementation? @lalitb ?
> > 
> > 
> > need to cosider version management: default will require all clients to update to 4.10, and vise versa in case of new version of okhttp, the library needs to be updated...
> 
> Got it, in that case, maybe this can be the default in the next major SDK version.

Yes, this sounds like good plan to give users option to migrate separately and test the implementation, and then make it default

**Pull request closed without merging**